node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/graph/DFAGraph.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.commons.path.fa.dfa.graph;
+
+import org.apache.iotdb.commons.path.fa.IFAState;
+import org.apache.iotdb.commons.path.fa.IFATransition;
+import org.apache.iotdb.commons.path.fa.dfa.DFAState;
+import org.apache.iotdb.commons.utils.TestOnly;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+public class DFAGraph {
+  private final List<IFAState> dfaStateList = new ArrayList<>();
+  private final List<IFAState>[] dfaTransitionTable;
+
+  public DFAGraph(NFAGraph nfaGraph, Collection<IFATransition> transitions) {
+    dfaTransitionTable = new List[transitions.size()];
+    int closureSize = nfaGraph.getStateSize();
+    // init start state
+    int index = 0;
+    // Map NFAStateClosure to DFASate
+    Map<Closure, DFAState> closureStateMap = new HashMap<>();
+    for (IFATransition transition : transitions) {
+      dfaTransitionTable[transition.getIndex()] = new ArrayList<>();
+      dfaTransitionTable[transition.getIndex()].add(null);
+    }
+    DFAState curState = new DFAState(index);
+    dfaStateList.add(curState);
+    Closure curClosure = new Closure(closureSize);
+    curClosure.addState(curState);
+    closureStateMap.put(curClosure, curState);
+    Stack<Closure> closureStack = new Stack<>();
+    closureStack.push(curClosure);
+
+    // construct DFA
+    while (!closureStack.isEmpty()) {
+      curClosure = closureStack.pop();
+      for (IFATransition transition : transitions) {
```

Comment:
Add annotation ```subset construction``` referred to fundamentals of compiling.

Reply:
done